### PR TITLE
Skip kube-dns tests if coredns is installed

### PR DESCRIPTION
**Pull request by upodroid**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
kops clusters use coredns instead of kube-dns and this test should not run if kube-dns is not installed.

#### Which issue(s) this PR fixes:

Part of #78995

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/autoscaling/dns_autoscaling.go [hunk 1]
```diff
@@ -50,15 +52,18 @@ var _ = SIGDescribe("DNS horizontal autoscaling", func() {
...
 	var DNSParams1 DNSParamsLinear
 	var DNSParams2 DNSParamsLinear
 	var DNSParams3 DNSParamsLinear
 
 	ginkgo.BeforeEach(func(ctx context.Context) {
-		e2eskipper.SkipUnlessProviderIs("gce", "gke")
 		c = f.ClientSet
+		dc = f.DynamicClient
+		gvr := schema.GroupVersionResource{Group: "apps", Version: "v1", Resource: "deployments/kube-dns"}
+		e2eskipper.SkipIfMissingResource(ctx, dc, gvr, metav1.NamespaceSystem)
```
Comment (aojea) [inline test/e2e/autoscaling/dns_autoscaling.go:66]: this method is not used anywhere AFAIK
```
grep -r SkipIfMissingResource test/
test/e2e/framework/skipper/skipper.go:// SkipIfMissingResource skips if the gvr resource is missing.
test/e2e/framework/skipper/skipper.go:func SkipIfMissingResource(ctx context.Context, dynamicClient dynamic.Interface, gvr schema.GroupVersionResource, namespace string) {
```
I prefer if you modify or duplicate `getDNSReplicas` to skip if the return value is 0

Comment (upodroid) [reply test/e2e/autoscaling/dns_autoscaling.go]: I reworked the testing logic

File: test/e2e/autoscaling/dns_autoscaling.go [hunk 2]
```diff
@@ -57,9 +57,13 @@ var _ = SIGDescribe("DNS horizontal autoscaling", func() {
 	var DNSParams3 DNSParamsLinear
 
 	ginkgo.BeforeEach(func(ctx context.Context) {
-		e2eskipper.SkipUnlessProviderIs("gce", "gke")
 		c = f.ClientSet
 
+		_, err := getDNSReplicas(ctx, c)
+		if err != nil && strings.Contains(err.Error(), "expected 1 DNS deployment") {
+			e2eskipper.Skipf(err.Error())
+		}
```
Comment (aojea) [inline test/e2e/autoscaling/dns_autoscaling.go:65]: just change the logic to return the error and the replicas and reuse the existing block in line 72

if err != nil {
   if strings.Contains(err.Error(), "expected 1 DNS deployment") && originDNSReplicasCount {
			e2eskipper.Skipf(err.Error())
   }
   framework.Failf("unexpcted error: %v", err)
```

```diff
diff --git a/test/e2e/autoscaling/dns_autoscaling.go b/test/e2e/autoscaling/dns_autoscaling.go
index 98f9945245f..563ba9a2ffe 100644
--- a/test/e2e/autoscaling/dns_autoscaling.go
+++ b/test/e2e/autoscaling/dns_autoscaling.go
@@ -311,7 +311,7 @@ func getDNSReplicas(ctx context.Context, c clientset.Interface) (int, error) {
                return 0, err
        }
        if len(deployments.Items) != 1 {
-               return 0, fmt.Errorf("expected 1 DNS deployment, got %v", len(deployments.Items))
+               return len(deployments.Items), fmt.Errorf("expected 1 DNS deployment, got %v", len(deployments.Items))
        }
 
        deployment := deployments.Items[0]
```

Comment (upodroid) [reply test/e2e/autoscaling/dns_autoscaling.go]: Fixed

**Comment by aojea** [comment]
/lgtm
/approve

Thanks

**Comment by upodroid** [comment]
:( 

Create a kind cluster and then check this value.

```
ubuntu@dev:~/go/src/k8s.io/kubernetes$ kubectl get deploy coredns -n kube-system -o yaml | yq .metadata
annotations:
  deployment.kubernetes.io/revision: "1"
creationTimestamp: "2023-09-19T11:36:37Z"
generation: 1
labels:
  k8s-app: kube-dns
name: coredns
namespace: kube-system
resourceVersion: "437"
uid: 6c6c2c66-fce5-4b54-b555-1849d32389b7
```

**Comment by upodroid** [comment]
I took a closer look and noticed the labels were present on the coredns manifests in cluster/* but not for kube-dns. I added them now.

**Pull request merged**